### PR TITLE
Stick pids

### PR DESCRIPTION
**Pull request by Quick-Flash**
I'm currently working on what should be a better version of cinematic setpoint. Basically I'm looking for feedback on how to make this better/simpler. The idea is to stick a sort of pid controller on the setpoint or more accurately the rcCommand (its a value of -500 to 500 based on your stick deflection). The point of this is to allow for a much more dynamic way of controlling stick feels. I have yet to test this, but I do feel like this needs some help and that I may be able to get some of that help here. 

I'm going to make a video discussing this code and what it means to do. I'll put that in here. Thanks for the help, I'm actually really excited about this and your feedback on this feature!
https://www.youtube.com/watch?v=OQLyKmTnVhA

### CODE REVIEW

File: src/main/fc/fc_rc.c [hunk 1]
```diff
@@ -854,3 +855,34 @@ bool rcSmoothingInitializationComplete(void) {
     return (rxConfig()->rc_smoothing_type != RC_SMOOTHING_TYPE_FILTER) || rcSmoothingData.filterInitialized;
 }
 #endif // USE_RC_SMOOTHING_FILTER
+
+FAST_CODE float stickFeels(float rcCommand, int axis)
+{
+  static FAST_RAM_ZERO_INIT float lastRcCommandData[3];
+  float pterm_low, pterm_high, pterm, iterm_low, iterm_high, iterm, dterm_low, dterm_high, dterm;
```
Comment (gretel) [inline src/main/fc/fc_rc.c]: Codacy Issue found: [The scope of the variable 'dterm' can be reduced.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5192242)

**Comment by gretel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5192242&bid=17031185)

File: src/main/fc/fc_rc.c [hunk 2]
```diff
@@ -854,3 +855,34 @@ bool rcSmoothingInitializationComplete(void) {
...
+  float rcCommandPercent;
+  float rcCommandError;
```
Comment (gretel) [inline src/main/fc/fc_rc.c]: Codacy Issue found: [The scope of the variable 'rcCommandError' can be reduced.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5192242)

**Comment by Quick-Flash** [comment]
just test flew this, it definitely works, but I still want a way for the center stick to have a bit more smooth at the end of a flip. Maybe i just need to tweak settings a bit different, idk

**Comment by Quick-Flash** [comment]
Just needs a bit more testing and I'll call it good. I'm liking it so far, just want a little bit more feedback.

**Comment by Quick-Flash** [comment]
I'm thinking that this is good to merge now!

**Comment by Quick-Flash** [comment]
It is ready :)

**Comment by gretel** [comment]
@Quick-Flash please resolve the conflict

File: src/main/build/version.h [hunk 3]
```diff
@@ -25,7 +25,7 @@
 #define FC_FIRMWARE_NAME            "EmuFlight"
 #define FC_VERSION_MAJOR            0  // increment when a major release is made (big new feature, etc)
 #define FC_VERSION_MINOR            2  // increment when a minor release is made (small new feature, change etc)
-#define FC_VERSION_PATCH_LEVEL      40 // increment when a bug is fixed
+#define FC_VERSION_PATCH_LEVEL      47 // increment when a bug is fixed
```
Comment (gretel) [inline src/main/build/version.h:28]: please start bumping version seperately cause its obstructive to work as a team

File: src/main/flight/pid.c [hunk 4]
```diff
@@ -96,21 +96,21 @@ PG_REGISTER_WITH_RESET_TEMPLATE(pidConfig_t, pidConfig, PG_PID_CONFIG, 2);
 #ifndef DEFAULT_PIDS_ROLL
 #define DEFAULT_PIDS_ROLL \
     {                     \
-        50, 70, 28, 0     \
+        50, 70, 28        \
```
Comment (gretel) [inline src/main/flight/pid.c:99]: superflous spaces please consider using editor-side formatting at least

File: src/main/flight/pid.h [hunk 5]
```diff
@@ -77,7 +75,7 @@ typedef struct pidf_s {
     uint8_t P;
     uint8_t I;
     uint8_t D;
-    uint16_t F;
+    uint16_t F; // Only keeping this for angle mode
```
Comment (gretel) [inline src/main/flight/pid.h:78]: why?

File: src/main/interface/msp.c [hunk 6]
```diff
@@ -1403,9 +1403,9 @@ bool mspProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst)
         sbufWriteU8(dst, 0);
 #endif
         sbufWriteU8(dst, 0);
-        sbufWriteU16(dst, currentPidProfile->pid[PID_ROLL].F);
-        sbufWriteU16(dst, currentPidProfile->pid[PID_PITCH].F);
-        sbufWriteU16(dst, currentPidProfile->pid[PID_YAW].F);
+        sbufWriteU16(dst, 0);
```
Comment (gretel) [inline src/main/interface/msp.c:1406]: wasnt this wrong like in another occurance?

File: src/main/sensors/battery.c [hunk 7]
```diff
@@ -116,7 +116,7 @@ PG_RESET_TEMPLATE(batteryConfig_t, batteryConfig,
     .vbatfullcellvoltage = 41,
 
     .vbatLpfPeriod = 35,
-    .ibatLpfPeriod = 40,
+    .ibatLpfPeriod = 10,
```
Comment (gretel) [inline src/main/sensors/battery.c:119]: this is not related

File: src/main/flight/pid.c (hunk 4 above)
Comment (Quick-Flash) [reply src/main/flight/pid.c:99]: not superflous if you look at the "\" character. I could the entire formatting or keep it this way. Either way i'm fine with.

File: src/main/flight/pid.h (hunk 5 above)
Comment (Quick-Flash) [reply src/main/flight/pid.h:78]: Angle mode still has access to an angle mode as the way angle mode is calculated is quite different from the regular pid controller.

File: src/main/interface/msp.c (hunk 6 above)
Comment (Quick-Flash) [reply src/main/interface/msp.c:1406]: I don't think it is, but i'll go and ask loutwice who is way better at the MSP to verify that it is working as expected.

File: src/main/flight/pid.c (hunk 4 above)
Comment (gretel) [reply src/main/flight/pid.c:99]: this comment was not related to the approval as the symbols next to it indicate

File: src/main/sensors/battery.c (hunk 7 above)
Comment (Quick-Flash) [reply src/main/sensors/battery.c:119]: yea i'll pull that in a separate pull request. thanks for showing me this

File: src/main/interface/msp.c (hunk 6 above)
Comment (gretel) [reply src/main/interface/msp.c:1406]: i guess he pushed another commit to fix something similar

File: src/main/flight/pid.h (hunk 5 above)
Comment (gretel) [reply src/main/flight/pid.h:78]: that make a better comment

File: src/main/build/version.h (hunk 3 above)
Comment (Quick-Flash) [reply src/main/build/version.h:28]: not entirely sure what you mean by this. I've been slowly bumping this as I have been testing it in order to easily understand what version testers have been using. I can make this 41 though if you want.

File: src/main/sensors/battery.c (hunk 7 above)
Comment (gretel) [reply src/main/sensors/battery.c:119]: this is what we do reviews for. like not getting another "telemetry" bug.

File: src/main/flight/pid.c (hunk 4 above)
Comment (Quick-Flash) [reply src/main/flight/pid.c:99]: i prefer the older style anyhow. i'll be putting it back to how it used to be.

File: src/main/build/version.h (hunk 3 above)
Comment (gretel) [reply src/main/build/version.h]: https://github.com/emuflight/EmuFlight/pull/119 introduced a less single-individual way of doing releases. you'd bump the version as part of a release process and keep the same version as long as the team... ya well.. decides on when to bump to the next version.
you work on one version until it is finished and just not bump each and any time.
without any roadmap it might be easy for you but it's not good for the project.

File: src/main/build/version.h [hunk 8]
```diff
@@ -25,7 +25,7 @@
 #define FC_FIRMWARE_NAME            "EmuFlight"
 #define FC_VERSION_MAJOR            0  // increment when a major release is made (big new feature, etc)
 #define FC_VERSION_MINOR            2  // increment when a minor release is made (small new feature, change etc)
-#define FC_VERSION_PATCH_LEVEL      47 // increment when a bug is fixed
+#define FC_VERSION_PATCH_LEVEL      41 // increment when a bug is fixed
```
Comment (gretel) [inline src/main/build/version.h:28]: nah now that should be 47 cause we cant get back in time can we

File: src/main/sensors/battery.c [hunk 9]
```diff
@@ -116,7 +116,7 @@ PG_RESET_TEMPLATE(batteryConfig_t, batteryConfig,
     .vbatfullcellvoltage = 41,
 
     .vbatLpfPeriod = 35,
-    .ibatLpfPeriod = 10,
+    .ibatLpfPeriod = 40,
```
Comment (gretel) [inline src/main/sensors/battery.c]: please check on https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history and learn how to interactively rebase so the change is not part of any commit

File: src/main/flight/pid.h (hunk 5 above)
Comment (Quick-Flash) [reply src/main/flight/pid.h:78]: woah just read my previous comment.. man it sounds dumb. What i meant to say is that the angle mode code is calculated quite differently than the normal acro code is. FeedForward is still being used in the angle pid controller. Proved not to be useful for acro, but angle FeedForward is slighly different for a few reasons.

File: src/main/build/version.h (hunk 8 above)
Comment (Quick-Flash) [reply src/main/build/version.h:28]: well we are going to go to 0.3.0 soon. We just got over one of the hurdles we had with the gui so a full release is looking a lot sooner. I can change it back to 2.47 though

File: src/main/flight/pid.h (hunk 5 above)
Comment (gretel) [reply src/main/flight/pid.h:78]: i was kidding in terms of put something in the code that explains it well please

File: src/main/build/version.h (hunk 8 above)
Comment (gretel) [reply src/main/build/version.h]: so we should work on version `0.3.0` *now* and release that to the public when done. like lots of other projects release just fine working on one version until its done and then proceed

**Comment by gretel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5192242&bid=17031185)

File: src/main/build/version.h [hunk 10]
```diff
@@ -25,7 +25,7 @@
 #define FC_FIRMWARE_NAME            "EmuFlight"
 #define FC_VERSION_MAJOR            0  // increment when a major release is made (big new feature, etc)
 #define FC_VERSION_MINOR            2  // increment when a minor release is made (small new feature, change etc)
-#define FC_VERSION_PATCH_LEVEL      47 // increment when a bug is fixed
+#define FC_VERSION_PATCH_LEVEL      48 // increment when a bug is fixed
```
Comment (gretel) [inline src/main/build/version.h:28]: this is still the same version. please do not bump

File: src/main/interface/msp.c (hunk 6 above)
Comment (Quick-Flash) [reply src/main/interface/msp.c]: Loutwice has seen this and it seems fine.

**Comment by gretel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5192242&bid=17031185)

File: src/main/build/version.h (hunk 10 above)
Comment (Quick-Flash) [reply src/main/build/version.h]: I just fixed this.

**Comment by gretel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5192242&bid=17031185)

**Comment by gretel** [comment]
this breaks `angle` mode badly. it never returns to center until first rc command pitching forward. some testing would be nice prior to merging changes.

**Comment by Quick-Flash** [comment]
it has been tested quite extensively. Project Mocking Bird has been flying both acro/horizon/angle on this code.

**Comment by gretel** [comment]
ah if  "project mockingbird" has been flying this it must be perfect. my reputation is no match. sorry

**Comment by madchiller** [comment]
No it just means it has been tested quite extensively.  I don’t understand why defensive and sarcastic.  Someone need a hug?

**Comment by loutwice** [comment]
well @gretel what is the behavior? it just keep the last input  until pitching forward?

**Comment by gretel** [comment]
@madchiller you were just what i was missing thank you so much

**Comment by gretel** [comment]
> well @gretel what is the behavior? it just keep the last input until pitching forward?

sort of aggregates the previous inputs to some new center for the angle mode. it's pretty strange and pretty dangerous for unexperienced pilots. but i guess i shouldnt talk about this much cause i have obviously triggered some bug drama.

**Comment by madchiller** [comment]
If your finding  a issue please bring it forward..    there is no drama I’m
aware of ..

Do you have any BBL of this issue?

**Comment by gretel** [comment]
@madchiller none of my tinyhawks has bbl. i dont run angle mode outdoors and currently i dont feel like trying.

**Comment by madchiller** [comment]
Is it possible it’s a tiny hawk issue??  Have you been able to see this on
a different setup ?

**Comment by gretel** [comment]
sorry i brought this up. no worries folks. i'm going to ask emax for a refund. wait it's working betaflight. am i being sarcastic again.

**Comment by Quick-Flash** [comment]
@gretel what settings are you using?
if you use correction end but not correction center then you can run into this as the correction does not ever disappear as technically the correction is an iterm...

**Comment by Quick-Flash** [comment]
so you would grow your correction then with no center correction you would have no way to get rid of the correction. in other words your rc command to the fc wouldn't be centered which can cause an issue like this.

**Comment by gretel** [comment]
> @gretel what settings are you using?
> if you use correction end but not correction center then you can run into this as the correction does not ever disappear as technically the correction is an iterm...

running the defaults. dont see any documentation on this. so i cant know what magic should be done to derive from the defaults so it's not.. well... a bug.

**Comment by Quick-Flash** [comment]
default values for this shouldn't do anything but i'll check the code again.

**Comment by gretel** [comment]
@Quick-Flash thanks

**Comment by gretel** [comment]
> @gretel what settings are you using?
> if you use correction end but not correction center then you can run into this as the correction does not ever disappear as technically the correction is an iterm...

meaing its extensively tested in a very limited range of parameters.. like requiring previous knowledge no to run into issues is like `boeing` does.

**Comment by Quick-Flash** [comment]
looking at things again. all the code does is change rcCommand. (fc_rc.c)
image

angle mode calculates the angle you want to be at using getRcDeflection. (pid.c)
image

getRcDeflection is really just a reflection of rcDeflection. (fc_rc.c)
image

rcDeflection is equal to rcCommandf. (fc_rc.c)
image

rcCommandf = rcCommand/500. (fc_rc.c)
image

So when calculating your angle of attack stick pids will effect what angle is desired. However, if you are running defaults then your sensitivity center and end will be 100, correction center and end will be 10 and weight center and end will both be 0. If it is set to this then stick pids actually has no effect. The code that changes your rcCommand (stick pids) will not run if sensitivity center and end are 100 and weight center and end are 0.

**Comment by gretel** [comment]
thanks for the explanation. so the defaults make this not having any effect?

**Comment by nerdCopter** [comment]
Defaults have no affect.
https://github.com/emuflight/EmuFlight/wiki/Rate-Dynamics

**Comment by gretel** [comment]
tried another tinyhawk (1 with `d8` rx) flashed with erase all no diff all defaults same behaviour.

**Comment by gretel** [comment]
"fixed" by downgrading to `0.2.0`. there is a regression for sure. repeating the "tested" paradigm will not help.

**Comment by madchiller** [comment]
How are we doing y’all? Jitter how are we confirming this issue? I’m trying
to see it in my Mob6 with angle mode and 2.49
But I’m not seeing it here...  flying seems normal and RX tab shows sticks
behavior correct...

**Comment by nerdCopter** [comment]
@Quick-Flash, you once explained that "some" correction is needed.  (And thus the default is `10`)
Should consideration be made to not allow <`10` or not allow `0` in both Firmware and Configurator?

**Comment by gretel** [comment]
tried with rate dynamics enabled, no change. tried today on 2.5" rig and angle mode is working fine. tinyhawks all `matekf411rx`
